README.md
# Foot Pedal Keyboard
A foot pedal keyboard, using Korg EC5, Arduino and Python
Converts pedal button events into operating system virtual keyboard sequences.

## Sample applications

### Media editing 📸
Control media editing software, such as Adobe Photosop and Premiere, with your foot

### Program code 👨‍💻
Use the foot keyboard pedal to send macros to a code IDE, such as Visual Studio Code

### Practicing musical instruments 🎸
Control playback with your foot, without stopping the practice. The hands stay on the instrument (e.g, Bass guitar) and hitting a foot pedal rewinds the playback to the start.

### Using with Practice#
For example: If Pedal A is clicked (pressed and released) then the python driver will generate a Media Prev keyboard stroke on the operating system. A musician can then control music playback by using his/her feet, a very natural way for musicians that use foot pedals for effects and volume control, when playing instruments such as guitar, bass guitar and keyboards.

Originally, I wanted to be able to able to control [Practice#](https://github.com/bigman73/practicesharp), my open source practice application, with my feet - instead of getting out of practice position and having to use my hands in order control the application 
With Practice# the setup is as followes:
+ Pedal A -> Look back (key = l)
+ Pedal B -> Rewind a few seconds back (key = ,)
+ Pedal C -> Toggle Play / Pause (key = space)
+ Pedal D -> Slow down (key = s)
+ Pedal E -> Speed up (key = f)

But I decided to make a more generic solution. The python driver supoorts dynamic supports any desktop application through an easy to configure json file (key-mappings.json). The default configuration file contains Chrome mappings - a tab can be closed with Pedal E!  

## Korg EC5 Foot Pedal Board
![alt text](https://raw.githubusercontent.com/bigman73/footpedalkeyboard/master/images/ec5-b.jpg "Korg EC5 Foot Pedal Board")

## The final result
![alt text](https://raw.githubusercontent.com/bigman73/footpedalkeyboard/master/images/pedal-final.jpg "Used in 'production' :)")
![alt text](https://raw.githubusercontent.com/bigman73/footpedalkeyboard/master/images/breadboard.jpg  "The breadboard solution")
![alt text](https://raw.githubusercontent.com/bigman73/footpedalkeyboard/master/images/python-driver-windows.png  "The FPK python driver running on Windows")

## More information - In the wiki
Bill Of Materials, wiring schematics and other details: 
https://github.com/bigman73/footpedalkeyboard/wiki

## Credits
I took some direction and ideas from a similar project done by Sylvain Wallez:
https://bluxte.net/musings/2016/07/29/usb-foot-keyboard/



footpedalkeyboard-driver/CHANGELOG.md
# Change Log

## Versions

1.2.0, 2020-11-28
- Code refactoring
- Code cleanup, with pylint
- Added mapping to VS Code
- Shortened driver python file to driver.py
- Updated conda dependencies

1.1.1, 2020-11-25
- Fixed typo
- Improved documentation

1.1.0, 2020-10-31
- Added support for process name based active application matching
- Fixed key mappings for Spotify to work on process name 
- Code refactor and cleanup

1.0.0, 2020-10-27
- Added key mappings support for Winamp 

footpedalkeyboard-driver/window_services.py
#!/usr/bin/env python

"""Find the currently active window."""

import logging
import sys

logging.basicConfig(format='%(asctime)s %(levelname)s %(message)s',
                    level=logging.DEBUG,
                    stream=sys.stdout)


def get_active_window():
    """
    Get the currently active window.

    Returns
    -------
    string :
        Name of the currently active window.
    """
    import sys
    active_window_name = None
    if sys.platform in ['linux', 'linux2']:
        # Alternatives: http://unix.stackexchange.com/q/38867/4784
        try:
            import wnck
        except ImportError:
            logging.info("wnck not installed")
            wnck = None
        if wnck is not None:
            screen = wnck.screen_get_default()
            screen.force_update()
            window = screen.get_active_window()
            if window is not None:
                pid = window.get_pid()
                with open("/proc/{pid}/cmdline".format(pid=pid)) as f:
                    active_window_name = f.read()
        else:
            try:
                from gi.repository import Gtk, Wnck
                gi = "Installed"
            except ImportError:
                logging.info("gi.repository not installed")
                gi = None
            if gi is not None:
                Gtk.init([])  # necessary if not using a Gtk.main() loop
                screen = Wnck.Screen.get_default()
                screen.force_update()  # recommended per Wnck documentation
                active_window = screen.get_active_window()
                pid = active_window.get_pid()
                with open("/proc/{pid}/cmdline".format(pid=pid)) as f:
                    active_window_name = f.read()
    elif sys.platform in ['Windows', 'win32', 'cygwin']:
        # http://stackoverflow.com/a/608814/562769
        import win32gui
        window = win32gui.GetForegroundWindow()
        active_window_name = win32gui.GetWindowText(window)
    elif sys.platform in ['Mac', 'darwin', 'os2', 'os2emx']:
        # http://stackoverflow.com/a/373310/562769
        from AppKit import NSWorkspace
        active_window_name = (NSWorkspace.sharedWorkspace()
                              .activeApplication()['NSApplicationName'])
    else:
        print("sys.platform={platform} is unknown. Please report."
              .format(platform=sys.platform))
        print(sys.version)
    return active_window_name

if __name__ == "__main__":
    print("Active window: %s" % str(get_active_window()))


footpedalkeyboard-driver/foodpedalkeyboard-driver.py
""" =============================================================================
MIT License

Copyright (c) 2018 Yuval Naveh (bigman73)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
=============================================================================== """

import sys
import serial
from serial.tools import list_ports
from pyautogui import press, keyDown, keyUp
import re
from window_services import get_active_window
import json

DRIVER_VERSION = "0.0.3"

# Teensy USB serial microcontroller program id data:
VENDOR_ID = "16C0"
PRODUCT_ID = "0483"
TEENSY_SIGNATURE = "USB VID:PID={0}:{1}".format(VENDOR_ID, PRODUCT_ID)

MAX_WINDOW_TITLE_CACHE_SIZE = 50

event_regex = re.compile('Pedal\s([A-E])\s(pressed|released)')

# Key: Mapping application name, Value: application mapping data
applications = {}

# Key: Window title, Value: Mapping application name
title_application_mapping_cache = {}


def find_teensy_port():
    print('Searching for a teensy COM port:')
    for port in list(list_ports.comports()):
        print(port, "=>", port[2])
        if port[2].startswith(TEENSY_SIGNATURE):
            return port[0]

    return None


def listen_to_teensy_serial(teensy_port):
    ser = serial.Serial(port=teensy_port, baudrate=9600, timeout=0.1)
    print('Listening to incoming serial messages...')
    print('-' * 80)

    while True:
        arduino_messsage = ser.readline()
        if arduino_messsage and len(arduino_messsage) > 0:
            event = arduino_messsage.decode("utf-8")
            event = event.replace('\r\n', '')
            process_footpedalkeyboard_event(event)
            print()


def process_footpedalkeyboard_event(event):
    print("RECEIVED: " + event)

    m = event_regex.match(event)
    if m:
        pedal = m.group(1)
        # print("[DEBUG] Pedal =", pedal)
        pedal_action = m.group(2)
        # print("[DEBUG] Action =", action)

        active_window_title = get_active_window()
        print("Active window:", active_window_title)

        matching_application = None

        cached_application_name = title_application_mapping_cache.get(active_window_title)

        if cached_application_name:
            print("Using a cached application =>", cached_application_name)
            matching_application = applications[cached_application_name]
        else:
            for application_name, application in applications.items():
                window_title_compiled_regex = application['compiled_regex']
                if window_title_compiled_regex.match(active_window_title):
                    print("Found a matching application =>", application_name)
                    matching_application = application
                    # Add to cache for next time
                    if title_application_mapping_cache.__len__() > MAX_WINDOW_TITLE_CACHE_SIZE:
                        title_application_mapping_cache.clear()
                    title_application_mapping_cache[active_window_title] = application['application_name']
                    break

        if matching_application:
            pedal_action_mappings = matching_application['mappings']['pedal' + pedal]
            if pedal_action_mappings:
                action_instructions = pedal_action_mappings.get(pedal_action)
                if action_instructions:
                    instruction_type = action_instructions['type']
                    if instruction_type == 'press':
                        keys = action_instructions['keys']
                        print("PRESS keys:", keys)
                        press(keys)
                    elif instruction_type == 'hotkey':
                        keys = action_instructions['keys']
                        print("Hot key:", keys)
                        for key in keys:
                            keyDown(key)
                        for key in keys:
                            keyUp(key)


def read_mapping_config_file():
    config_data = json.load(open('key-mappings.json'))

    for application_config in config_data['applications']:
        application_name = application_config['application_name']
        applications[application_name] = application_config
        window_title_regex = application_config['window_title_regex']
        application_config['compiled_regex'] = re.compile(window_title_regex)


if __name__ == "__main__":
    print('--- Foot Pedal Keyboard :: Driver, Version {0} ---'.format(DRIVER_VERSION))
    _teensy_port = find_teensy_port()
    if _teensy_port:
        print("GREAT SUCCESS: Teensy found on port {0}".format(_teensy_port))
        print()

        read_mapping_config_file()

        listen_to_teensy_serial(_teensy_port)
    else:
        print("No compatible Teensy found. Aborting.")

    sys.exit(1)


footpedalkeyboard-driver/footpedalkeyboard-driver.py
""" =============================================================================
MIT License

Copyright (c) 2018 Yuval Naveh (bigman73)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
=============================================================================== """

import sys
import serial
from serial.tools import list_ports
from pyautogui import press, keyDown, keyUp, typewrite
import re
from window_services import get_active_window
import json
import psutil
import win32process
import win32gui
import time

DRIVER_VERSION = "1.1.1"

# Teensy USB serial microcontroller program id data:
VENDOR_ID = "16C0"
PRODUCT_ID = "0483"
TEENSY_SIGNATURE = "USB VID:PID={0}:{1}".format(VENDOR_ID, PRODUCT_ID)

MAX_CACHE_SIZE = 50

event_regex = re.compile('Pedal\\s([A-E])\\s(pressed|released)')

# Key: Mapping application name, Value: application mapping data
applications = {}

# Key: Window title, Value: Mapping application name
title_application_mapping_cache = {}

# Key: Process Name, Value: Mapping application name
process_application_mapping_cache = {}


def get_active_window_process_name():
    """
    Gets the process name for the currently active window
    """

    # This produces a list of PIDs active window relates to
    pid = win32process.GetWindowThreadProcessId(win32gui.GetForegroundWindow())
    # pid[-1] is the most likely to survive last longer
    process_name = psutil.Process(pid[-1]).name()

    return process_name


def find_teensy_port():
    """
    Finds the teensy COM port
    """

    print('Searching for a teensy COM port:')
    for port in list(list_ports.comports()):
        print(port, "=>", port[2])
        if port[2].startswith(TEENSY_SIGNATURE):
            return port[0]

    return None


def listen_to_teensy_serial(teensy_port):
    """
    Listens to teensy serial messages
    """

    ser = serial.Serial(port=teensy_port, baudrate=9600, timeout=0.1)
    print('Listening to incoming serial messages...')
    print('-' * 80)

    while True:
        arduino_messsage = ser.readline()
        if arduino_messsage and len(arduino_messsage) > 0:
            event = arduino_messsage.decode("utf-8")
            event = event.replace('\r\n', '')
            process_footpedalkeyboard_event(event)
            print()


def process_footpedalkeyboard_event(event):
    """
    Processes an event coming from the foot pedal keyboard
    """

    print("RECEIVED KEYBOARD EVENT: " + event)

    matchingEvent = event_regex.match(event)
    if matchingEvent:
        pedal = matchingEvent.group(1)
        pedal_action = matchingEvent.group(2)

        active_window_title = get_active_window()
        print("Active window:", active_window_title)

        active_window_process_name = get_active_window_process_name()
        print("Active process:", active_window_process_name)

        matching_application = None

        cached_application_name = title_application_mapping_cache.get(
            active_window_title)

        # Lookup cache by active window title
        if cached_application_name:
            print("Using a cached application (by window title) =>",
                  cached_application_name)
            matching_application = applications[cached_application_name]

        # Lookup cache by process name of active window
        if not matching_application:
            cached_application_name = process_application_mapping_cache.get(
                active_window_process_name)
            if cached_application_name:
                print("Using a cached application (by process name) =>",
                      cached_application_name)
                matching_application = applications[cached_application_name]

        # Match application by active window's title
        if not matching_application:
            for application_name, application in applications.items():
                if not 'window_title_compiled_regex' in application:
                    continue

                window_title_compiled_regex = application['window_title_compiled_regex']
                if window_title_compiled_regex.match(active_window_title):
                    print("Found a matching application by window title =>",
                          application_name)
                    matching_application = application
                    # Add to cache for next time
                    if title_application_mapping_cache.__len__() > MAX_CACHE_SIZE:
                        title_application_mapping_cache.clear()
                    title_application_mapping_cache[active_window_title] = application['application_name']
                    break

        # Match application by active window's process name
        if not matching_application:
            for application_name, application in applications.items():
                if not 'process_name_compiled_regex' in application:
                    continue

                process_name_compiled_regex = application['process_name_compiled_regex']
                if process_name_compiled_regex.match(active_window_process_name):
                    print("Found a matching application by process name =>",
                          application_name)
                    matching_application = application
                    # Add to cache for next time
                    if process_application_mapping_cache.__len__() > MAX_CACHE_SIZE:
                        process_application_mapping_cache.clear()
                    process_application_mapping_cache[active_window_process_name] = application['application_name']
                    break

        if matching_application:
            apply_application_keys(matching_application, pedal, pedal_action)


def apply_application_keys(matching_application, pedal, pedal_action):
    """
    Applies the matching applications keys to the operating system
    """

    pedal_action_mappings = matching_application['mappings']['pedal' + pedal]
    if pedal_action_mappings:
        action_instructions = pedal_action_mappings.get(pedal_action)
        if action_instructions:
            instruction_type = action_instructions['type']
            if instruction_type == 'press':
                keys = action_instructions['keys']
                print("PRESS keys:", keys)
                press(keys)
            elif instruction_type == 'hotkey':
                keys = action_instructions['keys']
                print("Hot key:", keys)
                for key in keys:
                    keyDown(key)
                for key in keys:
                    keyUp(key)
            elif instruction_type == 'typewrite':
                text = action_instructions['text']
                interval = action_instructions['interval']
                typewrite(text, interval)


def read_mapping_config_file():
    """
    Reads the key mappings JSON configuration file
    """

    config_data = json.load(open('key-mappings.json'))

    all_applications_config = config_data['applications']
    print('Found {0} configured applications'.format(
        len(all_applications_config)))

    for application_config in all_applications_config:
        application_name = application_config['application_name']
        applications[application_name] = application_config

        if 'window_title_regex' in application_config:
            window_title_regex = application_config['window_title_regex']
            application_config['window_title_compiled_regex'] = re.compile(
                window_title_regex)

        if 'process_name_regex' in application_config:
            process_name_regex = application_config['process_name_regex']
            application_config['process_name_compiled_regex'] = re.compile(
                process_name_regex)


# Main program
if __name__ == "__main__":
    print(
        '--- Foot Pedal Keyboard :: Driver, Version {0} ---'.format(DRIVER_VERSION))
    _teensy_port = find_teensy_port()
    if _teensy_port:
        print("GREAT SUCCESS: Teensy found on port {0}".format(_teensy_port))
        print()

        read_mapping_config_file()

        listen_to_teensy_serial(_teensy_port)
    else:
        print("TOTAL FAIL NO WORKY: No compatible Teensy found. Aborting.")

    sys.exit(1)


footpedalkeyboard-driver/driver.py
""" =============================================================================
MIT License

Copyright (c) 2018 Yuval Naveh (bigman73)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
=============================================================================== """

import sys
import re
import json
import serial
from serial.tools import list_ports
from pyautogui import press, keyDown, keyUp, typewrite
import psutil
import win32process
import win32gui
from window_services import get_active_window

DRIVER_VERSION = "1.2.0"

# Teensy USB serial microcontroller program id data:
VENDOR_ID = "16C0"
PRODUCT_ID = "0483"
TEENSY_SIGNATURE = "USB VID:PID={0}:{1}".format(VENDOR_ID, PRODUCT_ID)

MAX_CACHE_SIZE = 50

event_regex = re.compile('Pedal\\s([A-E])\\s(pressed|released)')

# Key: Mapping application name, Value: application mapping data
applications = {}

# Key: Window title, Value: Mapping application name
title_application_mapping_cache = {}

# Key: Process Name, Value: Mapping application name
process_application_mapping_cache = {}


def get_active_window_process_name():
    """
    Gets the process name for the currently active window
    """

    # This produces a list of PIDs active window relates to
    pid = win32process.GetWindowThreadProcessId(win32gui.GetForegroundWindow())
    # pid[-1] is the most likely to survive last longer
    process_name = psutil.Process(pid[-1]).name()

    return process_name


def find_teensy_port():
    """
    Finds the teensy COM port
    """

    print('Searching for a teensy COM port:')
    for port in list(list_ports.comports()):
        print(port, "=>", port[2])
        if port[2].startswith(TEENSY_SIGNATURE):
            return port[0]

    return None


def listen_to_teensy_serial(teensy_port):
    """
    Listens to teensy serial messages
    """

    ser = serial.Serial(port=teensy_port, baudrate=9600, timeout=0.1)
    print('Listening to incoming serial messages...')
    print('-' * 80)

    while True:
        arduino_messsage = ser.readline()
        if arduino_messsage and len(arduino_messsage) > 0:
            event = arduino_messsage.decode("utf-8")
            event = event.replace('\r\n', '')
            process_footpedalkeyboard_event(event)
            print()


def match_application_by_window_title(window_title):
    """
    Finds a matching application by a given window title
    """

    matching_application = None

    for application_name, application in applications.items():
        # Only process applications that have a regular expression
        if not 'window_title_compiled_regex' in application:
            continue

        window_title_compiled_regex = application['window_title_compiled_regex']
        if window_title_compiled_regex.match(window_title):
            print("Found a matching application by window title =>",
                  application_name)
            matching_application = application

            # Add to cache for next time
            if title_application_mapping_cache.__len__() > MAX_CACHE_SIZE:
                title_application_mapping_cache.clear()
            title_application_mapping_cache[window_title] = application['application_name']
            break

    return matching_application


def match_application_by_process_name(window_process_name):
    """
       Finds a matching application by a given window process name
    """

    matching_application = None

    for application_name, application in applications.items():
        # Only process applications that have a regular expression
        if not 'process_name_compiled_regex' in application:
            continue

        process_name_compiled_regex = application['process_name_compiled_regex']
        if process_name_compiled_regex.match(window_process_name):
            print("Found a matching application by process name =>",
                  application_name)
            matching_application = application

            # Add to cache for next time
            if process_application_mapping_cache.__len__() > MAX_CACHE_SIZE:
                process_application_mapping_cache.clear()
            process_application_mapping_cache[window_process_name] = application['application_name']
            break

    return matching_application


def process_footpedalkeyboard_event(event):
    """
    Processes an event coming from the foot pedal keyboard
    """

    print("RECEIVED KEYBOARD EVENT: " + event)

    matching_event = event_regex.match(event)
    if matching_event:
        pedal = matching_event.group(1)
        pedal_action = matching_event.group(2)

        active_window_title = get_active_window()
        print("Active window:", active_window_title)

        active_window_process_name = get_active_window_process_name()
        print("Active process:", active_window_process_name)

        matching_application = None

        cached_application_name = title_application_mapping_cache.get(
            active_window_title)

        # Lookup cache by active window title
        if cached_application_name:
            print("Using a cached application (by window title) =>",
                  cached_application_name)
            matching_application = applications[cached_application_name]

        # Lookup cache by process name of active window
        if not matching_application:
            cached_application_name = process_application_mapping_cache.get(
                active_window_process_name)
            if cached_application_name:
                print("Using a cached application (by process name) =>",
                      cached_application_name)
                matching_application = applications[cached_application_name]

        # Match application by active window's title
        if not matching_application:
            matching_application = match_application_by_window_title(
                active_window_title)

        # Match application by active window's process name
        if not matching_application:
            matching_application = match_application_by_process_name(
                active_window_process_name)

        if matching_application:
            apply_application_keys(matching_application, pedal, pedal_action)


def apply_application_keys(matching_application, pedal, pedal_action):
    """
    Applies the matching applications keys to the operating system
    """

    pedal_action_mappings = matching_application['mappings']['pedal' + pedal]
    if pedal_action_mappings:
        action_instructions = pedal_action_mappings.get(pedal_action)
        if action_instructions:
            instruction_type = action_instructions['type']
            if instruction_type == 'press':
                keys = action_instructions['keys']
                print("PRESS keys:", keys)
                press(keys)
            elif instruction_type == 'hotkey':
                keys = action_instructions['keys']
                print("Hot key:", keys)
                for key in keys:
                    keyDown(key)
                for key in keys:
                    keyUp(key)
            elif instruction_type == 'typewrite':
                text = action_instructions['text']
                interval = action_instructions['interval']
                typewrite(text, interval)


def read_mapping_config_file():
    """
    Reads the key mappings JSON configuration file
    """

    config_data = json.load(open('key-mappings.json'))

    all_applications_config = config_data['applications']
    print('Found {0} configured applications'.format(
        len(all_applications_config)))

    for application_config in all_applications_config:
        application_name = application_config['application_name']
        applications[application_name] = application_config

        if 'window_title_regex' in application_config:
            window_title_regex = application_config['window_title_regex']
            application_config['window_title_compiled_regex'] = re.compile(
                window_title_regex)

        if 'process_name_regex' in application_config:
            process_name_regex = application_config['process_name_regex']
            application_config['process_name_compiled_regex'] = re.compile(
                process_name_regex)


# Main program
if __name__ == "__main__":
    print(
        '--- Foot Pedal Keyboard :: Driver, Version {0} ---'.format(DRIVER_VERSION))
    _teensy_port = find_teensy_port()
    if _teensy_port:
        print("GREAT SUCCESS: Teensy found on port {0}".format(_teensy_port))
        print()

        read_mapping_config_file()

        listen_to_teensy_serial(_teensy_port)
    else:
        print("TOTAL FAIL NO WORKY: No compatible Teensy found. Aborting.")

    sys.exit(1)
